shell/platform/fuchsia/flutter/engine.cc

```diff
@@ -54,14 +58,122 @@ std::unique_ptr<flutter::PlatformMessage> MakeLocalizationPlatformMessage(
-  return flutter::ThreadHost(name_prefix, flutter::ThreadHost::Type::RASTER |
-                                              flutter::ThreadHost::Type::UI |
-                                              flutter::ThreadHost::Type::IO);
+
+  // Default the config setter to setup the thread name only.
+  flutter::ThreadConfigSetter config_setter = fml::Thread::SetCurrentThreadName;
+
+  // Override the config setter if the media profile provider is available.
+  if (services) {
+    // Connect to the media profile provider to assign thread priorities using
+    // Fuchsia's scheduler role API. Failure to connect will print a warning and
+    // proceed with engine initialization, leaving threads created by the engine
+    // at default priority.
+    //
+    // The use of std::shared_ptr here is to work around the unfortunate
+    // requirement for flutter::ThreadConfigSetter (i.e. std::function<>) that
+    // the target callable be copy-constructible. This awkwardly conflicts with
+    // fuchsia::media::ProfileProviderSyncPtr being move-only. std::shared_ptr
+    // provides copyable object that references the move-only SyncPtr.
+    std::shared_ptr<fuchsia::media::ProfileProviderSyncPtr>
+        media_profile_provider =
+            std::make_shared<fuchsia::media::ProfileProviderSyncPtr>();
+
+    const zx_status_t connect_status =
+        services->Connect(media_profile_provider->NewRequest());
+    if (connect_status != ZX_OK) {
+      FML_LOG(WARNING)
+          << "Failed to connect to " << fuchsia::media::ProfileProvider::Name_
+          << ": " << zx_status_get_string(connect_status)
+          << " This is not a fatal error, but threads created by the engine "
+             "will run at default priority, regardless of the requested "
+             "priority.";
+    } else {
+      // This lambda must be copyable or the assignment fails to compile,
+      // necessitating the use of std::shared_ptr for the profile provider.
+      config_setter = [name_prefix, media_profile_provider =
+                                        std::move(media_profile_provider)](
+                          const fml::Thread::ThreadConfig& config) {
+        SetThreadConfig(name_prefix, *media_profile_provider, config);
+      };
+    }
+  }
+
+  flutter::ThreadHost::ThreadHostConfig thread_host_config{config_setter};
+
+  thread_host_config.SetRasterConfig(
+      {flutter::ThreadHost::ThreadHostConfig::MakeThreadName(
+           flutter::ThreadHost::Type::RASTER, name_prefix),
+       fml::Thread::ThreadPriority::RASTER});
+  thread_host_config.SetUIConfig(
+      {flutter::ThreadHost::ThreadHostConfig::MakeThreadName(
+           flutter::ThreadHost::Type::UI, name_prefix),
+       fml::Thread::ThreadPriority::DISPLAY});
+  thread_host_config.SetIOConfig(
+      {flutter::ThreadHost::ThreadHostConfig::MakeThreadName(
+           flutter::ThreadHost::Type::IO, name_prefix),
+       fml::Thread::ThreadPriority::DISPLAY});
```

Comment:
IO thread should be either ThreadPriority::NORMAL or ThreadPriority::BACKGROUND; it interacts with the disk.

Under the old logic, it was NORMAL (the default).  Probably wise to keep that to avoid unexpected regressions in asset loading performance.

Reply:
Looking at the codebase it looks like this abstraction won't be used for any ephemeral threads. Is that correct?

If so, then using NORMAL makes sense, since we can assign the corresponding role to whatever scheduling parameters make sense for the product requirements. If NORMAL were to be assigned to every new thread by default then we would have a problem, since IO latency matters on some products, and the IO thread would have to compete with every other random thread for CPU time.

Reply:
Changed the NORMAL based on this and offline conversation.